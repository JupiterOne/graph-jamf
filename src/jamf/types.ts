export interface Account {
  id: string;
  name: string;
}

export interface Site {
  id: number;
  name: string;
}

export type Category = Site;

export type PListDocument = string;

export interface AdminsAndGroups {
  users: Admin[];
  groups: Group[];
}

export interface Admin {
  id: number;
  name: string;
  directory_user?: boolean;
  full_name?: string;
  email?: string;
  email_address?: string;
  password_sha256?: string;
  enabled?: string;
  force_password_change?: boolean;
  access_level?: string;
  privilege_set?: string;
  privileges?: Privileges;
  groups?: {
    group: {
      id: number;
      name: string;
      site: Site;
      privileges: {
        jss_objects?: any;
        jss_settings?: any;
        recon?: any;
        casper_admin?: any;
        casper_remote?: any;
        casper_imaging?: any;
      };
    };
  };
}

export interface Group {
  id: number;
  name: string;
  site: Site;
  access_level?: string;
  privilege_set?: string;
  privileges?: Privileges;
  members?: Array<{ id: number; name: string }>;
}

export interface Privileges {
  jss_objects: string[];
  jss_settings?: string[];
  jss_actions?: string[];
  recon: string[];
  casper_admin?: string[];
  casper_remote: string[];
  casper_imaging: string[];
}

export interface User {
  id: number;
  name: string;
  full_name?: string;
  email?: string;
  email_address?: string;
  phone_number?: string;
  position?: string;
  enable_custom_photo_url?: boolean;
  custom_photo_url?: string;
  ldap_server?: {
    id: number;
    name: string;
  };
  extension_attributes?: Array<{
    extension_attribute: {
      id: number;
      name: string;
      type: string;
      value: string;
    };
  }>;
  sites?: Array<{
    site: Site;
  }>;
  links?: {
    computers: Array<{
      id: number;
      name: string;
    }>;
    peripherals: Array<{
      id: number;
      name: string;
    }>;
    mobile_devices: Array<{
      id: number;
      name: string;
    }>;
    vpp_assignments: Array<{
      id: number;
      name: string;
    }>;
    total_vpp_code_count: number;
  };
  managed_apple_id: string;
  user_groups: {
    size: number;
  };
}

export interface Application {
  name: string;
  path: string;
  version: string;
  bundle_id: string;
}

export interface ApplicationComputer {
  id: number;
  name: string;
  udid: string;
  serial_number: string;
  mac_address: string;
}

export interface ApplicationDetail {
  versions: Array<{
    number: string;
    computers: ApplicationComputer[];
  }>;
  unique_computers: ApplicationComputer[];
}

export interface DiskPartition {
  name: string;
  size: number;
  type: string;
  partition_capacity_mb: number;
  percentage_full: number;
  filevault_status: string;
  filevault_percent: number;
  filevault2_status: string;
  filevault2_percent: number;
  boot_drive_available_mb?: number;
  lvgUUID?: string;
  lvUUID?: string;
  pvUUID?: string;
  available_mb: number;
}

export interface Storage {
  disk: string;
  model: string;
  revision: string;
  serial_number: string;
  size: number;
  drive_capacity_mb: number;
  connection_type: string;
  smart_status: string;
  partition?: DiskPartition;
  partitions?: DiskPartition[];
}

export interface Printer {
  name: string;
  uri: string;
  type: string;
  location: string;
}

export interface Computer {
  id: number;
  name: string;
  managed: boolean;
  username: string;
  model: string;
  department: string;
  building: string;
  mac_address: string;
  udid: string;
  serial_number: string;
  report_date_utc: string;
  report_date_epoch: number;
}

export interface ComputerDetail {
  general: {
    id: number;
    name: string;
    network_adapter_type: string;
    alt_network_adapter_type: string;
    mac_address: string;
    alt_mac_address: string;
    ip_address: string;
    last_reported_ip: string;
    serial_number: string;
    udid: string;
    jamf_version: string;
    platform: string;
    barcode_1: string;
    barcode_2: string;
    asset_tag: string;
    supervised: boolean;
    remote_management: {
      managed: boolean;
      management_username: string;
      management_password_sha256: string;
    };
    mdm_capable: boolean;
    mdm_capable_users: { mdm_capable_user: string };
    mdm_profile_expiration_epoch: number;
    mdm_profile_expiration_utc: string;
    management_status: {
      enrolled_via_dep: boolean;
      user_approved_enrollment: boolean;
      user_approved_mdm: boolean;
    };
    report_date: string;
    report_date_epoch: number;
    report_date_utc: string;
    last_contact_time: string;
    last_contact_time_epoch: number;
    last_contact_time_utc: string;
    initial_entry_date: string;
    initial_entry_date_epoch: number;
    initial_entry_date_utc: string;
    last_cloud_backup_date_epoch: 0;
    last_cloud_backup_date_utc: string;
    last_enrolled_date_epoch: number;
    last_enrolled_date_utc: string;
    distribution_point: string;
    sus: string;
    netboot_server: string;
    site: Site;
    itunes_store_account_is_active: boolean;
  };
  location: {
    username: string;
    realname: string;
    real_name: string;
    email_address: string;
    position: string;
    phone: string;
    phone_number: string;
    department: string;
    building: string;
    room: string;
  };
  purchasing: {
    is_purchased: boolean;
    is_leased: boolean;
    po_number: string;
    vendor: string;
    applecare_id: string;
    purchase_price: string;
    purchasing_account: string;
    po_date: string;
    po_date_epoch: number;
    po_date_utc: string;
    warranty_expires: string;
    warranty_expires_epoch: number;
    warranty_expires_utc: string;
    lease_expires: string;
    lease_expires_epoch: number;
    lease_expires_utc: string;
    life_expectancy: number;
    purchasing_contact: string;
    os_applecare_id: string;
    os_maintenance_expires: string;
    attachments: string[];
  };
  // TODO: Add typings for this
  peripherals: any[];
  software: {
    unix_executables: any[];
    licensed_software: any[];
    installed_by_casper: any[];
    installed_by_installer_swu: string[];
    cached_by_casper: any[];
    available_software_updates: any[];
    available_updates: any;
    running_services: string[];
    applications: Application[];
    fonts: any[];
    plugins: any[];
  };
  extension_attributes: ExtensionAttribute[];
  hardware: {
    make: string;
    model: string;
    model_identifier: string;
    os_name: string;
    os_version: string;
    os_build: string;
    master_password_set: boolean;
    active_directory_status: string;
    service_pack: string;
    processor_type: string;
    processor_architecture: string;
    processor_speed: number;
    processor_speed_mhz: number;
    number_processors: number;
    number_cores: number;
    total_ram: number;
    total_ram_mb: number;
    boot_rom: string;
    bus_speed: number;
    bus_speed_mhz: number;
    battery_capacity: number;
    cache_size: number;
    cache_size_kb: number;
    available_ram_slots: number;
    optical_drive: string;
    nic_speed: string;
    smc_version: string;
    ble_capable: boolean;
    sip_status: string;
    gatekeeper_status: string;
    xprotect_version: string;
    institutional_recovery_key: string;
    disk_encryption_configuration: string;
    filevault2_users: string[];
    storage: Storage[];
    mapped_printers: Printer[];
  };
  certificates: {
    common_name: string;
    identity: boolean;
    expires_utc: string;
    expires_epoch: string | number;
    name: string;
  }[];
  security: {
    activation_lock: boolean;
    secure_boot_level: string;
    external_boot_level: string;
    firewall_enabled: boolean;
  };
  configuration_profiles: Array<{
    id: number;
    name: string;
    uuid: string;
    is_removable: boolean;
  }>;
  // TODO: Add typings for this
  groups_accounts: Record<string, any>;
  // TODO: Add typings for this
  iphones: any[];
}

export interface ExtensionAttribute {
  id: number;
  name: string;
  value: string[];
}

export interface Configuration {
  id: number;
  name: string;
}

export interface OSXConfigurationDetail {
  general: {
    id: number;
    name: string;
    description: string;
    site: Site;
    category: Category;
    distribution_method: string;
    user_removable: boolean;
    level: string;
    redeploy_on_update: string;
    payloads: PListDocument;
  };
  scope: {
    all_computers: boolean;
    all_jss_users: boolean;
  };
}

export interface OSXConfigurationPayload {
  PayloadDisplayName: string;
  PayloadContent: OSXConfigurationPayloadItem[];
}

export interface OSXConfigurationPayloadItem {
  PayloadType: string;
  PayloadEnabled: boolean;
  [key: string]: any;
}

export interface OSXConfigurationFirewallPayload
  extends OSXConfigurationPayloadItem {
  EnableFirewall: boolean;
  EnableStealthMode: boolean;
  BlockAllIncoming: boolean;
}

export interface OSXConfigurationScreensaverPayload
  extends OSXConfigurationPayloadItem {
  loginWindowIdleTime: number;
}

export interface OSXConfigurationDetailParsed extends OSXConfigurationDetail {
  parsedPayload: OSXConfigurationPayload;
}

export interface MobileDevice {
  id: number;
  name: string;
  device_name: string;
  udid: string;
  serial_number: string;
  phone_number: string;
  wifi_mac_address: string;
  managed: boolean;
  supervised: boolean;
  model: string;
  model_identifier: string;
  model_display: string;
  username: string | undefined;
}

export enum Method {
  GET = 'get',
  POST = 'post',
}

export interface AccountsResponse {
  accounts: AdminsAndGroups;
}

export interface AdminResponse {
  account: Admin;
}

export interface GroupResponse {
  group: Group;
}

export interface UsersResponse {
  users: User[];
}

export interface UserResponse {
  user: User;
}

export interface ApplicationDetailResponse {
  computer_applications: ApplicationDetail;
}

export interface ComputerResponse {
  computers: Computer[];
}

export interface ComputerDetailResponse {
  computer: ComputerDetail;
}

export interface OSXConfigurationResponse {
  os_x_configuration_profiles: Configuration[];
}

export interface OSXConfigurationDetailResponse {
  os_x_configuration_profile: OSXConfigurationDetail;
}

export interface MobileDevicesResponse {
  mobile_devices: MobileDevice[];
}

export interface JamfDataModel {
  users: User[];
  mobileDevices: MobileDevice[];
  computers: Computer[];
  computerDetails: ComputerDetail[];
  osxConfigurationDetails: OSXConfigurationDetailParsed[];
  admins: Admin[];
  groups: Group[];

  osxConfigurationDetailsById: DataByID<OSXConfigurationDetailParsed>;
}

export interface DataByID<T> {
  [id: number]: T;
}
