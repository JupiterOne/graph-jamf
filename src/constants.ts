export const IngestionSources = {
  ACCOUNTS: 'accounts',
  MACOS_CONFIGURATION_PROFILES: 'macos-configuration-profiles',
  MOBILE_DEVICES: 'mobile-devices',
  COMPUTERS: 'computers',
  GROUPS: 'groups',
  ADMINS: 'admins',
};
