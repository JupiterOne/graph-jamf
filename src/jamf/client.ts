import fetch, { RequestInit, Response, FetchError } from 'node-fetch';
import { retry } from '@lifeomic/attempt';
import PQueue from 'p-queue';
import {
  Admin,
  AdminsAndGroups,
  ApplicationDetail,
  Computer,
  ComputerDetail,
  Configuration,
  Group,
  Method,
  MobileDevice,
  OSXConfigurationDetail,
  User,
  AccountsResponse,
  AdminResponse,
  ApplicationDetailResponse,
  ComputerDetailResponse,
  ComputerResponse,
  GroupResponse,
  MobileDevicesResponse,
  OSXConfigurationDetailResponse,
  OSXConfigurationResponse,
  UserResponse,
  UsersResponse,
} from './types';
import {
  IntegrationLogger,
  IntegrationProviderAPIError,
} from '@jupiterone/integration-sdk-core';
import { URL } from 'url';
import { IntegrationProviderPermanentAPIError } from '../util/error';

interface OnApiRequestErrorParams {
  url: string;
  err: FetchError;
  attemptNum: number;
  attemptsRemaining: number;
  code?: string;
}

type RequestFunction = (
  url: string,
  options?: RequestInit | undefined,
) => Promise<Response>;

interface CreateJamfClientParams {
  host: string;
  username: string;
  password: string;
  request?: RequestFunction;
  onApiRequestError?: (params: OnApiRequestErrorParams) => void;
}

const defaultApiTimeoutMs = 60000; // 1 minute
const noRetryStatusCodes: number[] = [400, 401, 403, 404, 413];

function isSuccessfulStatusCode(status: number) {
  return status >= 200 && status < 400;
}

async function request(
  requestFn: RequestFunction,
  url: string,
  options?: RequestInit | undefined,
): Promise<Response> {
  const response = await requestFn(url, options);

  if (isSuccessfulStatusCode(response.status)) {
    return response;
  }

  if (noRetryStatusCodes.includes(response.status)) {
    throw new IntegrationProviderPermanentAPIError({
      endpoint: url,
      statusText: 'Received non-retryable status code in API response',
      status: response.status,
    });
  }

  throw new IntegrationProviderAPIError({
    endpoint: url,
    statusText: response.statusText,
    status: response.status,
  });
}

export class JamfClient {
  private readonly queue: PQueue;
  private readonly host: string;
  private readonly username: string;
  private readonly password: string;
  private readonly request: RequestFunction;

  private readonly onApiRequestError:
    | ((params: OnApiRequestErrorParams) => void)
    | undefined;

  constructor(options: CreateJamfClientParams) {
    this.host = options.host;
    this.username = options.username;
    this.password = options.password;
    this.request = options.request || fetch;
    this.onApiRequestError = options.onApiRequestError;

    this.queue = new PQueue({ concurrency: 1, intervalCap: 1, interval: 50 });
  }

  /**
   * Requires "Read - Jamf Pro User Accounts & Groups" permission
   */
  public async fetchAccounts(): Promise<AdminsAndGroups> {
    const result = await this.makeRequest<AccountsResponse>(
      `/accounts`,
      Method.GET,
      {},
    );
    const { users, groups } = result.accounts;

    return { users, groups };
  }

  /**
   * Requires "Read - Jamf Pro User Accounts & Groups" permission
   */
  public async fetchAccountUserById(id: number): Promise<Admin> {
    const result = await this.makeRequest<AdminResponse>(
      `/accounts/userid/${id}`,
      Method.GET,
      {},
    );

    return result.account;
  }

  /**
   * Requires "Read - Jamf Pro User Accounts & Groups" permission
   */
  public async fetchAccountGroupById(id: number): Promise<Group> {
    const result = await this.makeRequest<GroupResponse>(
      `/accounts/groupid/${id}`,
      Method.GET,
      {},
    );

    return result.group;
  }

  /**
   * Requires "Read - Users" permission
   */
  public async fetchUsers(): Promise<User[]> {
    const result = await this.makeRequest<UsersResponse>(
      `/users`,
      Method.GET,
      {},
    );

    return result.users;
  }

  /**
   * Requires "Read - Users" permission
   */
  public async fetchUserById(id: number): Promise<User> {
    const result = await this.makeRequest<UserResponse>(
      `/users/id/${id}`,
      Method.GET,
      {},
    );

    return result.user;
  }

  /**
   * Requires "Read - Mobile Devices" permission
   */
  public async fetchMobileDevices(): Promise<MobileDevice[]> {
    const result = await this.makeRequest<MobileDevicesResponse>(
      '/mobiledevices',
      Method.GET,
      {},
    );

    return result.mobile_devices;
  }

  /**
   * Requires "Read - Computers" permission
   */
  public async fetchComputers(): Promise<Computer[]> {
    const result = await this.makeRequest<ComputerResponse>(
      '/computers/subset/basic',
      Method.GET,
      {},
    );

    return result.computers;
  }

  /**
   * Requires "Read - Computers" permission
   */
  public async fetchComputerById(id: number): Promise<ComputerDetail> {
    const result = await this.makeRequest<ComputerDetailResponse>(
      `/computers/id/${id}`,
      Method.GET,
      {},
    );

    return result.computer;
  }

  /**
   * Requires "Read - Advanced Computer Searches" permission
   */
  public async fetchComputerApplicationByName(
    name: string,
  ): Promise<ApplicationDetail> {
    const result = await this.makeRequest<ApplicationDetailResponse>(
      `/computerapplications/application/${name}`,
      Method.GET,
      {},
    );

    return result.computer_applications;
  }

  /**
   * Requires "Read - macOS Configuration Profiles" permission
   */
  public async fetchOSXConfigurationProfiles(): Promise<Configuration[]> {
    const result = await this.makeRequest<OSXConfigurationResponse>(
      `/osxconfigurationprofiles`,
      Method.GET,
      {},
    );

    return result.os_x_configuration_profiles;
  }

  /**
   * Requires "Read - macOS Configuration Profiles" permission
   */
  public async fetchOSXConfigurationProfileById(
    id: number,
  ): Promise<OSXConfigurationDetail> {
    const result = await this.makeRequest<OSXConfigurationDetailResponse>(
      `/osxconfigurationprofiles/id/${id}`,
      Method.GET,
      {},
    );

    return result.os_x_configuration_profile;
  }

  public getResourceUrl(path: string): string {
    const url = new URL(
      this.host.startsWith('http') ? this.host : `https://${this.host}`,
    );
    url.pathname = `JSSResource${path}`;
    return url.href;
  }

  private async makeRequest<T>(
    path: string,
    method: Method,
    params: {},
    headers?: {},
  ): Promise<T> {
    const options: RequestInit = {
      method,
      timeout: defaultApiTimeoutMs,
      headers: {
        'Content-Type': 'application/json',
        Accept: 'application/json',
        Authorization: `Basic ${Buffer.from(
          this.username + ':' + this.password,
        ).toString('base64')}`,
        'Accept-Encoding': 'identity',
        ...headers,
      },
    };

    const fullUrl = this.getResourceUrl(path);

    // The goal here is to retry and ensure the final error includes information
    // about the host we could not connect to, since users define the host and
    // may mis-type the value.
    const requestWithRetry = (): Promise<Response> =>
      retry(async () => request(this.request, fullUrl, options), {
        maxAttempts: 5,
        handleError: (err, attemptContext) => {
          if (err.retryable === false) {
            attemptContext.abort();
            return;
          }

          const fetchErr = err as FetchError;

          if (attemptContext.attemptsRemaining && this.onApiRequestError) {
            // If there are no attempts remaining, we will just bubble up the
            // entire error by default.
            this.onApiRequestError({
              url: fullUrl,
              code: fetchErr.code,
              err,
              attemptNum: attemptContext.attemptNum,
              attemptsRemaining: attemptContext.attemptsRemaining,
            });
          }
        },
      });

    const response = await this.queue.add(requestWithRetry);

    if (response.status === 200) {
      return response.json();
    } else {
      throw new IntegrationProviderAPIError({
        endpoint: fullUrl,
        statusText: response.statusText,
        status: response.status,
      });
    }
  }
}

interface CreateJamfClientHelperParams {
  host: string;
  username: string;
  password: string;
  logger: IntegrationLogger;
}

export function createClient({
  host,
  username,
  password,
  logger,
}: CreateJamfClientHelperParams) {
  const client = new JamfClient({
    host,
    username,
    password,
    onApiRequestError(requestError) {
      logger.info(requestError, 'Error making API requests (will retry)');
    },
  });

  return client;
}
