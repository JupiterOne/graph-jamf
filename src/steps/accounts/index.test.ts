import { fetchAccounts } from '.';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { AdminsAndGroups } from '../../jamf/types';
import { ACCOUNT_DATA_KEY, Entities } from '../constants';

describe('#fetchAccounts', () => {
  test('should collect data', async () => {
    const { context } = await createDataCollectionTest({
      recordingName: 'fetchAccounts',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [fetchAccounts],
      entitySchemaMatchers: [
        {
          _type: Entities.ACCOUNT._type,
          matcher: {
            _class: ['Account'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'jamf_account' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
              },
            },
          },
        },
      ],
    });

    const accountData =
      await context.jobState.getData<AdminsAndGroups>(ACCOUNT_DATA_KEY);

    expect(accountData).toMatchSnapshot('accountData');
  });
});
