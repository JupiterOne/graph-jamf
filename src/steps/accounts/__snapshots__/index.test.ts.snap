// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchAccounts should collect data: accountData 1`] = `
{
  "groups": [
    {
      "id": 1,
      "name": "Test Group",
      "site": {
        "id": 1,
        "name": "Test site",
      },
    },
    {
      "id": 2,
      "name": "Another Test Group",
      "site": {
        "id": 1,
        "name": "Test site",
      },
    },
    {
      "id": 3,
      "name": "Not all groups have members inside in api",
      "site": {
        "id": 1,
        "name": "Test site",
      },
    },
  ],
  "users": [
    {
      "id": 12,
      "name": "adam.pierson",
    },
    {
      "id": 1,
      "name": "adam.williams",
    },
    {
      "id": 9,
      "name": "austin.kelleher",
    },
    {
      "id": 13,
      "name": "chasen.bettinger",
    },
    {
      "id": 3,
      "name": "erkang.zheng",
    },
    {
      "id": 7,
      "name": "isaacwilliams",
    },
    {
      "id": 8,
      "name": "jupiterone.dev",
    },
    {
      "id": 14,
      "name": "max.daum",
    },
    {
      "id": 17,
      "name": "my.new.account",
    },
    {
      "id": 10,
      "name": "nick.dowmon",
    },
    {
      "id": 15,
      "name": "samuel.poulton",
    },
    {
      "id": 11,
      "name": "stefan.virag",
    },
    {
      "id": 16,
      "name": "Wrong_Permissions",
    },
  ],
}
`;

exports[`#fetchAccounts should collect data: jobState 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Account",
      ],
      "_key": "jamf_account_local-integration-instance",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "id": "local-integration-instance",
            "name": "Local Integration",
          },
        },
      ],
      "_type": "jamf_account",
      "createdOn": undefined,
      "displayName": "Local Integration",
      "id": "local-integration-instance",
      "name": "Local Integration",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "jamf_account",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
